pkg/database/authorized_app.go

```diff
@@ -229,16 +227,35 @@ func (db *Database) SaveAuthorizedApp(r *AuthorizedApp) error {
 	return db.db.Save(r).Error
 }
 
-// hmacAPIKey is a helper for generating the HMAC of an API key. It returns the
-// hex-encoded HMACed value, suitable for insertion into the database.
-func (db *Database) hmacAPIKey(v string) (string, error) {
-	sig := hmac.New(sha512.New, db.config.APIKeyDatabaseHMAC)
-	if _, err := sig.Write([]byte(v)); err != nil {
-		return "", nil
+// GenerateAPIKeyHMAC generates the HMAC of the provided API key using the
+// latest HMAC key.
+func (db *Database) GenerateAPIKeyHMAC(apiKey string) (string, error) {
+	keys := db.config.APIKeyDatabaseHMAC
+	if len(keys) < 1 {
+		return "", fmt.Errorf("expected at least 1 hmac key")
+	}
+
+	sig := hmac.New(sha512.New, keys[0])
+	if _, err := sig.Write([]byte(apiKey)); err != nil {
+		return "", err
 	}
 	return base64.RawURLEncoding.EncodeToString(sig.Sum(nil)), nil
 }
 
+// generateAPIKeyHMACs creates a permutation of all possible API keys based on
+// the provided HMACs. It's primarily used to find an API key in the database.
+func (db *Database) generateAPIKeyHMACs(apiKey string) ([]string, error) {
```

Comment:
why is one public but not the other?

Reply:
Because we never want someone outside the package to be able to generate the _full_ HMAC set (which is only used for comparisons). Outside of the package, consumers should only generate the _latest_ HMAC.